### DISCUSSION TITLE
How to use arrays in Companion / with 'split' ?

### DISCUSSION THREAD

**Discussion by Coupe70** [bitfocus/companion#2989] [Q&A] [answered]
Hi,

it seems like you can use arrays in Companion, but I couldn't find any documentation how to do this.
There is this hint
![grafik](https://github.com/user-attachments/assets/34bec521-b346-4766-b8ec-dd90a4e4f5f2)
and functions like
![grafik](https://github.com/user-attachments/assets/8f547bf8-c99f-44b0-b25b-0cf9c0df251c)
that should probably split a string into an array and wouldn't make much sense without arrays.

I tried to give a custom variable the value [1, 2, 3, 4], but seems to result in a string "[1, 2, 3, 4]"
being the value of the variable. I also tried to set a custom variable to split("good#morning#world", "#"),
but this results in "good,morning,world" being the value of the variable. So yes, it splits, but doesn't
result in an array...

And if it worked, how to access one element of the array?
For now $(internal:custom_testarray)[1] just results in "[1, 2, 3, 4][1]" for the first example.
$(internal:custom_testarray[1]) doesn't work either.

Any idea how to get this working?

**Comment by Julusian** [comment, marked as answer]
Only fields which have 'expression' in their label support the expression syntax, other fields will at best support substituting variables.
So I expect that when you are getting `[1, 2, 3, 4][1]`, you are doing this in a field which doesn't support expressions.

The place where you create custom-variables also doesn't support expressions, that input field sets the custom variable to a string.
I dont remember the full rules on how well custom-variables support being an array, but I think they do.

**Comment by Coupe70** [comment]
Oh yes, I got it!
I thought it would be a bit more versatile and I tried to display `$(internal:custom_testarray)[1]` as button text.
`$(internal:custom_testarray)` works as it is a variable, `$(internal:custom_testarray)[1]` does not work as it is considered 
an expression. 

When I use 'internal: Custom Variable: Set with expression' on my variable and use [1, 2, 3, 4] as expression it results
in 1,2,3,4 in the variable - so there is some "array conversion" happening and this seems to be the way an "array"
is looking in Companion.
I can also use 'internal: Custom Variable: Set with expression' on a different variable and set it to $(internal:custom_testarray)[2].
The result is 2 which would be correct for a one-based array.

But: If I use `$(internal:custom_testarray)[1]` I get a comma (,).
I tried further and `$(internal:custom_testarray)[x]` seems to be nothig else than what `StringMid(var, x, 1)` is in other programming
languages, delivering just the one sign at position x. Time for a bug report it seems...

Anyway, thanks for pointing me in the right direction!
